pkg/remotewrite/client.go

```diff
@@ -0,0 +1,129 @@
+package remotewrite
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/sirupsen/logrus"
+
+	"github.com/pyroscope-io/pyroscope/pkg/config"
+	"github.com/pyroscope-io/pyroscope/pkg/convert/pprof"
+	"github.com/pyroscope-io/pyroscope/pkg/ingestion"
+	"github.com/pyroscope-io/pyroscope/pkg/storage/segment"
+)
+
+var (
+	ErrConvertPutInputToRequest = errors.New("failed to convert putInput into a http.Request")
+	ErrMakingRequest            = errors.New("failed to make request")
+	ErrNotOkResponse            = errors.New("response not ok")
+)
+
+type Client struct {
+	log    *logrus.Logger
+	config config.RemoteWrite
+	client *http.Client
+}
+
+func NewClient(logger *logrus.Logger, cfg config.RemoteWrite) *Client {
+	client := &http.Client{
+		// TODO(eh-am): make timeout configurable
+		Timeout: time.Second * 15,
+	}
+
+	return &Client{
+		log:    logger,
+		config: cfg,
+		client: client,
+	}
+}
+
+func (r *Client) Ingest(ctx context.Context, in *ingestion.IngestInput) error {
+	req, err := r.ingestInputToRequest(in)
+	if err != nil {
+		return multierror.Append(err, ErrConvertPutInputToRequest)
+	}
+
+	r.enhanceWithAuth(req)
+
+	req = req.WithContext(ctx)
+	r.log.Debugf("Making request to %s", req.URL.String())
+	res, err := r.client.Do(req)
+	if err != nil {
+		return multierror.Append(err, ErrMakingRequest)
+	}
+
+	if !(res.StatusCode >= 200 && res.StatusCode < 300) {
+		// read all the response body
+		respBody, _ := ioutil.ReadAll(res.Body)
+		return multierror.Append(ErrNotOkResponse, fmt.Errorf("status code: '%d'. body: '%s'", res.StatusCode, respBody))
+	}
+
+	return nil
```

Comment:
I think we need to close response body to allow Transport to reuse the connection:
```
// The http Client and Transport guarantee that Body is always
// non-nil, even on responses without a body or responses with
// a zero-length body. It is the caller's responsibility to
// close Body. The default HTTP client's Transport may not
// reuse HTTP/1.x "keep-alive" TCP connections if the Body is
// not read to completion and closed.
```
https://cs.opensource.google/go/go/+/go1.18.3:src/net/http/response.go;l=35